group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupMetadataManager.java

```diff
@@ -874,4 +1045,1265 @@ public void replay(
             consumerGroup.updateMember(newMember);
         }
     }
+
+    // Below stores all methods to handle generic group APIs.
+
+    /**
+     * Handle a JoinGroupRequest.
+     *
+     * @param context The request context.
+     * @param request The actual JoinGroup request.
+     *
+     * @return The result that contains records to append if the join group phase completes.
+     */
+    public CoordinatorResult<CompletableFuture<Errors>, Record> genericGroupJoin(
+        RequestContext context,
+        JoinGroupRequestData request,
+        CompletableFuture<JoinGroupResponseData> responseFuture
+    ) {
+        CoordinatorResult<CompletableFuture<Errors>, Record> result = EMPTY_RESULT;
+        String groupId = request.groupId();
+        String memberId = request.memberId();
+        int sessionTimeoutMs = request.sessionTimeoutMs();
+
+        if (sessionTimeoutMs < groupMinSessionTimeoutMs ||
+            sessionTimeoutMs > groupMaxSessionTimeoutMs
+        ) {
+            responseFuture.complete(new JoinGroupResponseData()
+                .setMemberId(memberId)
+                .setErrorCode(Errors.INVALID_SESSION_TIMEOUT.code())
+            );
+        } else {
+            boolean isUnknownMember = memberId.equals(UNKNOWN_MEMBER_ID);
+            // Group is created if it does not exist and the member id is UNKNOWN. if member
+            // is specified but group does not exist, request is rejected with GROUP_ID_NOT_FOUND
+            GenericGroup group;
+            try {
+                group = (GenericGroup) getOrMaybeCreateGroup(groupId, GENERIC, isUnknownMember);
```

Comment:
i don't think this fits well because we may have other records to append in this method. As mentioned above, I will store the generic groups in a separate hash map to prevent this from happening.